lambda/translcomprehension.ml

```diff
@@ -2,304 +2,406 @@ open Lambda
-    | Guarded (bdy_and_len, arr_kind, id) ->
-      let body = Ident.create_local "body" in
-      let len = Ident.create_local "len" in
-      Lstaticcatch(bdy_and_len, (id, [(body, Pgenval); (len, Pintval)]),
-        Lsequence(
-            Lsequence(
-              Lprim(Parraysets(arr_kind),
-                [Lvar(arr); Lvar(counter); Lvar(body)], loc),
-              Lprim(Parraysets(arr_kind),
-                [Lvar(arr); Lprim(Paddint, [Lvar(counter); (int 1)], loc);
-                Lvar(len)], loc)),
-          Lassign(counter, Lprim(Paddint, [Lvar(counter); (int 2)], loc)))),
-      None, arr_kind
+(* Generate code to initialise an element of an "uninitialised" array *)
+let init_array_elem ~loc ~kind ~size ~array ~index ~value =
+  let set_elem =
+    Lprim(Parraysetu kind, [Lvar array; Lvar index; Lvar value], loc)
   in
-  let body, arrs =
-    match guard with
-      | None ->
-        body, Array_of_arrays(arrs)
-      | Some guard ->
-        Lifthenelse(
-          (transl_exp ~scopes guard),
-          body,
-          lambda_unit
-        ), Array_of_filtered_arrays(arrs)
+  match kind with
+  | Pgenarray ->
+      let is_first_iteration =
+        Lprim(Pintcomp Ceq, [Lvar index; int 0], loc)
+      in
+      let make_array =
+        Lassign(array, make_array_prim ~loc size (Lvar value))
+      in
+      Lifthenelse(is_first_iteration, make_array, set_elem)
+  | Pintarray | Paddrarray | Pfloatarray -> set_elem
+
+(* Generate code to blit elements into an "uninitialised" array *)
+let init_array_elems ~loc ~kind ~size ~array ~index ~src ~len =
+  let blit =
+    blit_array_prim ~loc (Lvar src) (int 0) (Lvar array) (Lvar index) (Lvar len)
   in
-  let body, lengths, materialize =
-    transl_loops  ~transl_exp  ~loc ~scopes clauses body in
-  let block_len = Ident.create_local "block_len" in
-  let body, len =
-    match global_counter, comp_block with
-    | Some gc, Unguarded _ ->
-      Lsequence(
-        body,
-        Lassign(gc, Lprim(Paddint, [Lvar(counter); Lvar(gc)], loc))),
-      Lvar(block_len)
-    | Some gc, Guarded _ ->
+  match kind with
+  | Pgenarray ->
+      let is_first_iteration =
+        Lprim(Pintcomp Ceq, [Lvar index; int 0], loc)
+      in
+      let is_not_empty =
+        Lprim(Pintcomp(Cne), [Lvar len; int 0], loc)
+      in
+      let first_elem =
+        Lprim(Parrayrefu kind, [Lvar src; int 0], loc)
+      in
+      let make_array =
+        Lassign(array, make_array_prim ~loc size first_elem)
+      in
       Lsequence(
-        body,
-        Lassign(gc, Lprim(Paddint,
-          [Lvar(gc);
-          Lprim(Pdivint(Unsafe), [Lvar(counter); int 2], loc)], loc))),
-      Lprim(Pmulint, [Lvar(block_len); int 2], loc)
-    | None, Unguarded _  -> body, Lvar(block_len)
-    | None, Guarded _  -> body, Lprim(Pmulint, [Lvar(block_len); int 2], loc)
+        Lifthenelse(is_first_iteration,
+          Lifthenelse(is_not_empty,
+                      Lsequence(make_array, blit),
```

Comment:
```suggestion
                      make_array,
```

This blits twice in the first iteration.

The other possible change is to replace the second lambda_unit with blit and remove the outermost Lsequence.

Reply:
Good catch. Fixed.